### ISSUE TITLE
`AutoEscape::None` depends on the calling order from version 1.0.11

### ISSUE THREAD

**Issue by smihica** [mitsuhiko/minijinja#401]
Hi. Thank you for the great software.
We have hit upon an unexpected behavior.

## Description

Our code was like:

```rust
use minijinja::{context, AutoEscape, Environment};

fn main() {
    let mut env = Environment::new();
    env.add_template("hello.html", "<html><body>{{ body }}</body></html>")
        .unwrap();
    env.set_auto_escape_callback(|_| AutoEscape::None);
    println!(
        "{}",
        env.get_template("hello.html")
            .unwrap()
            .render(context! { body => "<p>hello</p>" })
            .unwrap()
    );
}
```

Until 1.0.10, it worked as we expected.

```
$ cargo run
<html><body><p>hello</p></body></html>
```

But after I updated minijinja to 1.0.11, it broke up. it seems same in 1.0.12.

```
$ cargo run
<html><body>&lt;p&gt;hello&lt;&#x2f;p&gt;</body></html>
```

I tried to change the order

```rust
env.set_auto_escape_callback(|_| AutoEscape::None);
env.add_template("hello.html", "<html><body>{{ body }}</body></html>")
    .unwrap();
```

and it started working in all versions.

```
$ cargo run
<html><body><p>hello</p></body></html>
```

Is it intended?
Does the `env.set_auto_escape_callback(|_| AutoEscape::None);` depend on calling order?

Thank you!

**Comment by mitsuhiko** [comment]
It's both intended and unintended. The behavior change was not really considered however the previous behavior of inheriting the auto escape callback at template load time had some unexpected consequences and meant that the template reference needed to hold on to that, rather than the compile template. The change landed as part of #384

What is intentional here is that the template reference when looked up from `get_template` does not have to hold on to anything other than the reference to the environment and the reference to the internal template state. However implementation wise the consequence of this is that this information is now attached at the time one calls `add_template` which is not great. However that said, this is not the only information that is attached to the template this way.  For instance `add_template` also has to access the configuration so even in older versions flipping some changes applied to `add_template` like the current syntax. This in a way just "fixes" that inconsistency where only some values had this behavior.

That said, this is clearly a change where the consequence was not considered and the documentation is also misleading as it states that the function is called when loading "from" the environment. In all reality the behavior has always depended on if a loader was used or `add_template` was called in that regard.

At the very least the documentation needs to be updated, but also it should be considered if this is a problem that should be reverted.

**Comment by smihica** [comment]
Thank you for the comment.
In our project, we just changed the ordering so both are ok.

But It might be better to document ^^
